<a href="https://colab.research.google.com/github/Pavanvarma007/-Pavan_INFO5731_Spring2023/blob/main/in_class_exercise_04_pavanvarma_info5731.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **The fourth in-class-exercise (40 points in total, 03/28/2022)**

Question description: Please use the text corpus you collected in your last in-class-exercise for this exercise. Perform the following tasks:

## (1) (10 points) Generate K topics by using LDA, the number of topics K should be decided by the coherence score, then summarize what are the topics. You may refer the code here: 

https://www.machinelearningplus.com/nlp/topic-modeling-gensim-python/

In [ ]:
# Importing modules
import pandas as pd
from google.colab import drive
drive.mount('/content/gdrive')

import pandas as pd

papers = pd.read_csv('papers.csv')
papers.head()



Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


id  year                                              title event_type  \
0     1  1987  Self-Organization of Associative Database and ...        NaN   
1    10  1987  A Mean Field Theory of Layer IV of Visual Cort...        NaN   
2   100  1988  Storing Covariance by the Associative Long-Ter...        NaN   
3  1000  1994  Bayesian Query Construction for Neural Network...        NaN   
4  1001  1994  Neural Network Ensembles, Cross Validation, an...        NaN   

                                            pdf_name          abstract  \
0  1-self-organization-of-associative-database-an...  Abstract Missing   
1  10-a-mean-field-theory-of-layer-iv-of-visual-c...  Abstract Missing   
2  100-storing-covariance-by-the-associative-long...  Abstract Missing   
3  1000-bayesian-query-construction-for-neural-ne...  Abstract Missing   
4  1001-neural-network-ensembles-cross-validation...  Abstract Missing   

                                          paper_text  
0  767\n\nSELF-ORGANIZATION OF ASSOCIATIVE DATABA...  
1  683\n\nA MEAN FIELD THEORY OF LAYER IV OF VISU...  
2  394\n\nSTORING COVARIANCE BY THE ASSOCIATIVE\n...  
3  Bayesian Query Construction for Neural\nNetwor...  
4  Neural Network Ensembles, Cross\nValidation, a...

In [ ]:
# Remove the columns
papers = papers.drop(columns=['id', 'event_type', 'pdf_name'], axis=1).sample(100)

papers.head()

year                                              title  \
440   1997  Synchronized Auditory and Cognitive 40 Hz Atte...   
4844  2014  Fundamental Limits of Online and Distributed A...   
6814  2017  Permutation-based Causal Inference Algorithms ...   
2241  2006    Modeling Dyadic Data with Binary Latent Factors   
2262  2006  The Robustness-Performance Tradeoff in Markov ...   

                                               abstract  \
440                                    Abstract Missing   
4844  Many machine learning approaches are character...   
6814  Learning directed acyclic graphs using both ob...   
2241                                   Abstract Missing   
2262                                   Abstract Missing   

                                             paper_text  
440   Synchronized Auditory and Cognitive 40 Hz\nAtt...  
4844  Fundamental Limits of Online and Distributed\n...  
6814  Permutation-based Causal Inference Algorithms\...  
2241  Modeling Dyadic Data with Binary Latent Factor...  
2262  The Robustness-Performance Tradeoff in Markov\...

In [ ]:
# Load the regular expression library
import re
# Remove punctuation
papers['paper_text_processed'] = \
papers['paper_text'].map(lambda x: re.sub('[,\.!?]', '', x))
# Convert the titles to lowercase
papers['paper_text_processed'] = \
papers['paper_text_processed'].map(lambda x: x.lower())
# Print out the first rows of papers
papers['paper_text_processed'].head()

440     synchronized auditory and cognitive 40 hz\natt...
4844    fundamental limits of online and distributed\n...
6814    permutation-based causal inference algorithms\...
2241    modeling dyadic data with binary latent factor...
2262    the robustness-performance tradeoff in markov\...
Name: paper_text_processed, dtype: object

In [ ]:
import gensim
from gensim.utils import simple_preprocess
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
stop_words = stopwords.words('english')
stop_words.extend(['from', 'subject', 're', 'edu', 'use'])
def sent_to_words(sentences):
    for sentence in sentences:
        # deacc=True removes punctuations
        yield(gensim.utils.simple_preprocess(str(sentence), deacc=True))
def remove_stopwords(texts):
    return [[word for word in simple_preprocess(str(doc)) 
             if word not in stop_words] for doc in texts]
data = papers.paper_text_processed.values.tolist()
data_words = list(sent_to_words(data))
# remove stop words
data_words = remove_stopwords(data_words)
print(data_words[:1][0][:30])

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


['synchronized', 'auditory', 'cognitive', 'hz', 'attentional', 'streams', 'impact', 'rhythmic', 'expectation', 'auditory', 'scene', 'analysis', 'bill', 'baird', 'dept', 'mathematics', 'ucberkeley', 'berkeley', 'ca', 'baird', 'mathberkeleyedu', 'abstract', 'developed', 'neural', 'network', 'architecture', 'implements', 'theory', 'attention', 'learning']


In [ ]:
import gensim.corpora as corpora
# Create Dictionary
id2word = corpora.Dictionary(data_words)
# Create Corpus
texts = data_words
# Term Document Frequency
corpus = [id2word.doc2bow(text) for text in texts]
# View
print(corpus[:1][0][:30])

[(0, 1), (1, 1), (2, 1), (3, 8), (4, 1), (5, 1), (6, 1), (7, 1), (8, 1), (9, 1), (10, 1), (11, 1), (12, 1), (13, 1), (14, 3), (15, 3), (16, 3), (17, 1), (18, 2), (19, 2), (20, 1), (21, 2), (22, 1), (23, 1), (24, 29), (25, 1), (26, 1), (27, 2), (28, 1), (29, 1)]


In [ ]:
from pprint import pprint
# number of topics
num_topics = 10
# Build LDA model
lda_model = gensim.models.LdaMulticore(corpus=corpus,
                                       id2word=id2word,
                                       num_topics=num_topics)
# Print the Keyword in the 10 topics
pprint(lda_model.print_topics())
doc_lda = lda_model[corpus]

[(0,
  '0.006*"learning" + 0.005*"model" + 0.005*"data" + 0.004*"one" + 0.004*"set" '
  '+ 0.004*"time" + 0.004*"algorithm" + 0.003*"function" + 0.003*"models" + '
  '0.003*"using"'),
 (1,
  '0.005*"learning" + 0.004*"data" + 0.004*"model" + 0.004*"set" + '
  '0.004*"using" + 0.004*"function" + 0.004*"input" + 0.003*"figure" + '
  '0.003*"time" + 0.003*"problem"'),
 (2,
  '0.006*"learning" + 0.005*"data" + 0.005*"network" + 0.004*"model" + '
  '0.004*"matrix" + 0.004*"time" + 0.004*"algorithm" + 0.004*"one" + '
  '0.003*"also" + 0.003*"set"'),
 (3,
  '0.006*"learning" + 0.006*"model" + 0.004*"time" + 0.004*"models" + '
  '0.004*"one" + 0.003*"using" + 0.003*"used" + 0.003*"algorithm" + '
  '0.003*"two" + 0.003*"data"'),
 (4,
  '0.005*"model" + 0.005*"data" + 0.005*"learning" + 0.004*"time" + '
  '0.004*"matrix" + 0.003*"algorithm" + 0.003*"used" + 0.003*"also" + '
  '0.003*"log" + 0.003*"using"'),
 (5,
  '0.005*"learning" + 0.005*"matrix" + 0.005*"model" + 0.004*"algorithm" + '
  '0.00

In [ ]:
pip install pyLDAvis

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 55.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 298.0/298.0 KB 29.6 MB/s eta 0:00:00
  Attempting uninstall: joblib
    Found existing installation: joblib 1.1.1
    Uninstalling joblib-1.1.1:
      Successfully uninstalled joblib-1.1.1
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pandas-profiling 3.2.0 requires joblib~=1.1.0, but you have joblib 1.2.0 which is incompatible.


In [ ]:
#import pyLDAvis.gensim

import pyLDAvis
import pyLDAvis.gensim_models as gensimvis
pyLDAvis.enable_notebook()


LDAvis_prepared = pyLDAvis.gensim_models.prepare(lda_model, corpus, id2word)

pyLDAvis.display(LDAvis_prepared)

/usr/local/lib/python3.9/dist-packages/pyLDAvis/_prepare.py:243: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  default_term_info = default_term_info.sort_values(


## (2) (10 points) Generate K topics by using LSA, the number of topics K should be decided by the coherence score, then summarize what are the topics. You may refer the code here:

https://www.datacamp.com/community/tutorials/discovering-hidden-topics-python

In [ ]:
# Write your code here

import pandas as pd

# load data
df = pd.read_csv('Musical_instruments_reviews.csv', usecols=['reviewerID', 'reviewText'])
df.head()





/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


reviewerID                                         reviewText
0  A2IBPI20UZIR0U  Not much to write about here, but it does exac...
1  A14VAT5EAX3D9S  The product does exactly as it should and is q...
2  A195EZSQDW3E21  The primary job of this device is to block the...
3  A2C00NNG1ZQQG2  Nice windscreen protects my MXL mic and preven...
4   A94QU4C90B1AX  This pop filter is great. It looks and perform...

In [ ]:
print(df.shape)
df = df.dropna()
print(df.shape)

(10261, 2)
(10254, 2)


/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
from gensim.parsing.preprocessing import remove_stopwords, strip_punctuation \
                                        , preprocess_string, strip_short, stem_text

# preprocess given text
def preprocess(text):
    
    # clean text based on given filters
    CUSTOM_FILTERS = [lambda x: x.lower(), 
                                remove_stopwords, 
                                strip_punctuation, 
                                strip_short, 
                                stem_text]
    text = preprocess_string(text, CUSTOM_FILTERS)
    
    return text

# apply function to all reviews 
df['Text (Clean)'] = df['reviewText'].apply(lambda x: preprocess(x))

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
from gensim import corpora

# create a dictionary with the corpus
corpus = df['Text (Clean)']
dictionary = corpora.Dictionary(corpus)

# convert corpus into a bag of words
bow = [dictionary.doc2bow(text) for text in corpus]

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
from gensim.models import LsiModel
from gensim.models.coherencemodel import CoherenceModel

# find the coherence score with a different number of topics
for i in range(2,11):
    lsi = LsiModel(bow, num_topics=i, id2word=dictionary)
    coherence_model = CoherenceModel(model=lsi, texts=df['Text (Clean)'], dictionary=dictionary, coherence='c_v')
    coherence_score = coherence_model.get_coherence()
    print('Coherence score with {} clusters: {}'.format(i, coherence_score))

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning: Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.
  sparsetools.csc_matvecs(


Coherence score with 2 clusters: 0.4638705176288849


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning: Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.
  sparsetools.csc_matvecs(


Coherence score with 3 clusters: 0.45170263620829987


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning: Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.
  sparsetools.csc_matvecs(


Coherence score with 4 clusters: 0.3782478735195774


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning: Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.
  sparsetools.csc_matvecs(


Coherence score with 5 clusters: 0.43493372415976417


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning: Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.
  sparsetools.csc_matvecs(


Coherence score with 6 clusters: 0.4380524171610878


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning: Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.
  sparsetools.csc_matvecs(


Coherence score with 7 clusters: 0.4093608603227179


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning: Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.
  sparsetools.csc_matvecs(


Coherence score with 8 clusters: 0.3785383655136472


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning: Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.
  sparsetools.csc_matvecs(


Coherence score with 9 clusters: 0.38800194062554416


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning: Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.
  sparsetools.csc_matvecs(


Coherence score with 10 clusters: 0.3611579064738203


In [ ]:
lsi = LsiModel(bow, num_topics=10, id2word=dictionary)

for topic_num, words in lsi.print_topics(num_words=5):
    print('Words in {}: {}.'.format(topic_num, words))

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning: Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.
  sparsetools.csc_matvecs(


Words in 0: 0.329*"sound" + 0.314*"guitar" + 0.242*"string" + 0.232*"pedal" + 0.217*"amp".
Words in 1: -0.584*"string" + 0.428*"pedal" + -0.380*"guitar" + 0.312*"amp" + -0.161*"tune".
Words in 2: 0.529*"string" + 0.367*"pedal" + -0.261*"mic" + -0.190*"record" + 0.187*"tone".
Words in 3: 0.641*"pedal" + -0.520*"amp" + -0.306*"sound" + 0.178*"guitar" + -0.119*"mic".
Words in 4: 0.641*"guitar" + 0.351*"amp" + -0.342*"string" + -0.269*"mic" + -0.192*"record".
Words in 5: 0.460*"tuner" + -0.382*"sound" + 0.371*"tune" + 0.309*"amp" + -0.246*"guitar".
Words in 6: -0.848*"pick" + 0.214*"guitar" + -0.172*"plai" + -0.155*"like" + 0.119*"sound".
Words in 7: 0.426*"sound" + -0.384*"strap" + 0.358*"tuner" + -0.350*"cabl" + 0.308*"tune".
Words in 8: -0.450*"cabl" + 0.436*"stand" + 0.399*"mic" + -0.324*"sound" + -0.230*"good".
Words in 9: -0.671*"strap" + 0.354*"cabl" + 0.273*"guitar" + -0.181*"sound" + -0.181*"like".


## (3) (10 points) Generate K topics by using  lda2vec, the number of topics K should be decided by the coherence score, then summarize what are the topics. You may refer the code here:

https://nbviewer.org/github/cemoody/lda2vec/blob/master/examples/twenty_newsgroups/lda2vec/lda2vec.ipynb

In [ ]:
nltk.download('all')
!pip install preprocess
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline






In [ ]:
import pyLDAvis
pyLDAvis.enable_notebook()

In [ ]:
top = 10
topic_to_topwords = {}
for j, topic in enumerate(table_review.Review.tolist()):
    top1 = np.argsort(topic)[::-1][:top]
    msg = 'Topic %i '  % j
    tp_wrd = [table_review.Review.tolist()[i].strip()[:35] for i in top1]
    msg += ' has top words '.join(tp_wrd)
    print(msg)
    topic_to_topwords[j] = tp_wrd

In [ ]:
plt.bar(np.arange(20), table_review.Review.tolist()[1])

## (4) (10 points) Generate K topics by using BERTopic, the number of topics K should be decided by the coherence score, then summarize what are the topics. You may refer the code here: 

https://colab.research.google.com/drive/1FieRA9fLdkQEGDIMYl0I3MCjSUKVF8C-?usp=sharing

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')



/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

df = pd.read_csv('/content/tokyo_2020_tweets.csv')
df = df.dropna()
df = df.sample(20000, random_state=42)
df.head()

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
<ipython-input-18-cf17a970fe83>:5: DtypeWarning: Columns (0,1,2,3,4,5,6,7,8,9,10,11,12,15) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/content/tokyo_2020_tweets.csv')


id                      user_name  \
295003  1419957058938486804  Australia in Belgium 🇦🇺🇪🇺🇧🇪🇱🇺   
63678   1419307203933229065                   念真 Nian Zhen   
101824  1419592988833947650            Indonesia Badminton   
60985   1419311331782447108                    FirstSportz   
47225   1419235863100370946           Kurdistan 24 English   

               user_location  \
295003     Brussels, Belgium   
63678   Enshi, Hubei, China.   
101824             Indonesia   
60985              New Delhi   
47225       Erbil, Kurdistan   

                                         user_description  \
295003  Australian Mission to the European Union and N...   
63678   一个中国\n\nChina institute of contemporary intern...   
101824  Indonesia Badminton Supporter • Click to follo...   
60985   Official Twitter handle of Firstsportz. \nFoll...   
47225   The English service of Kurdistan's leading new...   

               user_created user_followers user_friends user_favourites  \
295003  2014-11-26 15:52:22           3766         3226            4622   
63678   2017-06-30 20:07:29           7788         87.0         38102.0   
101824  2009-11-07 23:26:53          55711          451            2081   
60985   2019-11-26 14:14:55            478        258.0          4053.0   
47225   2015-07-06 08:32:11         139803          5.0             7.0   

       user_verified                 date  \
295003          True  2021-07-27 09:45:19   
63678          False  2021-07-25 14:43:01   
101824         False  2021-07-26 09:38:37   
60985          False  2021-07-25 14:59:25   
47225           True  2021-07-25 09:59:32   

                                                     text  \
295003  Congrats to 🇦🇺, 🇧🇪, and 🇱🇺 for our achievement...   
63678   Reality            Vs      New York Times\n#To...   
101824  [Live Score] #Tokyo2020   \n\nWD Group A : \n\...   
60985   Tennis at Tokyo Olympics: Stefanos #Tsitsipas ...   
47225   The 41-year-old gold medalist is an ethnic Kur...   

                         hashtags               source  retweets  favorites  \
295003              ['Tokyo2020']      Twitter Web App       0.0        0.0   
63678               ['Tokyo2020']  Twitter for Android      12.0       57.0   
101824              ['Tokyo2020']   Twitter for iPhone       0.0        4.0   
60985               ['Tsitsipas']  Twitter for Android       0.0        0.0   
47225   ['Tokyo2020', 'Olympics']      Twitter Web App       3.0        9.0   

       is_retweet  
295003      False  
63678       False  
101824      False  
60985       False  
47225       False

In [ ]:
import re
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

nltk.download('stopwords')
nltk.download('punkt')

stop_words = stopwords.words('english')

def clean_text(x):
  x = str(x)
  x = x.lower()
  x = re.sub(r'#[A-Za-z0-9]*', ' ', x)
  x = re.sub(r'https*://.*', ' ', x)
  x = re.sub(r'@[A-Za-z0-9]+', ' ', x)
  tokens = word_tokenize(x)
  x = ' '.join([w for w in tokens if not w.lower() in stop_words])
  x = re.sub(r'[%s]' % re.escape('!"#$%&\()*+,-./:;<=>?@[\\]^_`{|}~“…”’'), ' ', x)
  x = re.sub(r'\d+', ' ', x)
  x = re.sub(r'\n+', ' ', x)
  x = re.sub(r'\s{2,}', ' ', x)
  return x


df['clean_text'] = df.text.apply(clean_text)
df.head()

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
<>:19: DeprecationWarning: invalid escape sequence \(
<>:19: DeprecationWarning: invalid escape sequence \(
<ipython-input-19-ee8d035ba5d1>:19: DeprecationWarning: invalid escape sequence \(
  x = re.sub(r'[%s]' % re.escape('!"#$%&\()*+,-./:;<=>?@[\\]^_`{|}~“…”’'), ' ', x)
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


id                      user_name  \
295003  1419957058938486804  Australia in Belgium 🇦🇺🇪🇺🇧🇪🇱🇺   
63678   1419307203933229065                   念真 Nian Zhen   
101824  1419592988833947650            Indonesia Badminton   
60985   1419311331782447108                    FirstSportz   
47225   1419235863100370946           Kurdistan 24 English   

               user_location  \
295003     Brussels, Belgium   
63678   Enshi, Hubei, China.   
101824             Indonesia   
60985              New Delhi   
47225       Erbil, Kurdistan   

                                         user_description  \
295003  Australian Mission to the European Union and N...   
63678   一个中国\n\nChina institute of contemporary intern...   
101824  Indonesia Badminton Supporter • Click to follo...   
60985   Official Twitter handle of Firstsportz. \nFoll...   
47225   The English service of Kurdistan's leading new...   

               user_created user_followers user_friends user_favourites  \
295003  2014-11-26 15:52:22           3766         3226            4622   
63678   2017-06-30 20:07:29           7788         87.0         38102.0   
101824  2009-11-07 23:26:53          55711          451            2081   
60985   2019-11-26 14:14:55            478        258.0          4053.0   
47225   2015-07-06 08:32:11         139803          5.0             7.0   

       user_verified                 date  \
295003          True  2021-07-27 09:45:19   
63678          False  2021-07-25 14:43:01   
101824         False  2021-07-26 09:38:37   
60985          False  2021-07-25 14:59:25   
47225           True  2021-07-25 09:59:32   

                                                     text  \
295003  Congrats to 🇦🇺, 🇧🇪, and 🇱🇺 for our achievement...   
63678   Reality            Vs      New York Times\n#To...   
101824  [Live Score] #Tokyo2020   \n\nWD Group A : \n\...   
60985   Tennis at Tokyo Olympics: Stefanos #Tsitsipas ...   
47225   The 41-year-old gold medalist is an ethnic Kur...   

                         hashtags               source  retweets  favorites  \
295003              ['Tokyo2020']      Twitter Web App       0.0        0.0   
63678               ['Tokyo2020']  Twitter for Android      12.0       57.0   
101824              ['Tokyo2020']   Twitter for iPhone       0.0        4.0   
60985               ['Tsitsipas']  Twitter for Android       0.0        0.0   
47225   ['Tokyo2020', 'Olympics']      Twitter Web App       3.0        9.0   

       is_retweet                                         clean_text  
295003      False    congrats 🇦🇺 🇧🇪 🇱🇺 achievements far olympics 👏👏👏  
63678       False                          reality vs new york times  
101824      False   live score wd group greysia polii apriyani ra...  
60985       False  tennis tokyo olympics stefanos becomes first g...  
47225       False   year old gold medalist ethnic kurd western ir...

In [ ]:
!pip install bertopic

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 120.7/120.7 KB 9.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.2/5.2 MB 70.9 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.2/88.2 KB 8.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 KB 8.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.8/6.8 MB 39.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 48.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 199.8/199.8 KB 19.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 45.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     

In [ ]:
from bertopic import BERTopic
tweets = df.clean_text.to_list()
timestamp = df.date.to_list()

topic_model = BERTopic(language="english")
topics, probs = topic_model.fit_transform(tweets)

topic_model.get_topic_info()

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Topic  Count                                     Name
0       -1   6489              -1_athletes_day_tokyo_match
1        0    497              0_medals_medal_bronze_tally
2        1    396        1_olympics_olympic_games_athletes
3        2    246  2_weightlifting_weightlifter_mirabai_kg
4        3    181           3_swimming_swim_swimmers_swims
..     ...    ...                                      ...
346    345     10              345_barty_gone_ash_huurrhhh
347    346     10          346_germany_apet_hooter_germans
348    347     10          347_swims_canada_warn_vancouver
349    348     10            348_il_brisa_matisse_hennessy
350    349     10           349_flag_raising_bearer_jayson

[351 rows x 3 columns]

In [ ]:
topic_model.get_topic_info(0)

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Topic  Count                         Name
0      0    497  0_medals_medal_bronze_tally

In [ ]:
topic_model.visualize_topics()

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
topic_model.visualize_barchart(top_n_topics=12, n_words = 10, width = 350, height = 350)

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



## (5) (10 extra points) Compare the results generated by the four topic modeling algorithms, which one is better? You should explain the reasons in details.

# Write your answer here (no code needed for this question)
The results of LDA are as follows: '0.007"data" + 0.006"learning" + 0.005"model" + 0.005"algorithm" + ' '0.004"using" + 0.004"time" + 0.004"set" + 0.003"two" + 0.003"neural" + ' '0.003"function"

The results of LSA are as follows: 0.329"sound" + 0.314"guitar" + 0.242"string" + 0.232"pedal" + 0.217*"amp".

The results of lda2vec are as follows: topic 00 which has top words x11r5, xv, window, xterm, server, motif, font, xlib, //, sunos

The results of Bertopic are as follows: olympics_olympic_games_athletes_olympians_watching sports_summer

Bertopic results are more easily interpretable.